lib/sendf.c

```diff
@@ -328,9 +328,15 @@ CURLcode Curl_write(struct connectdata *conn,
   bytes_written = conn->send[num](conn, num, mem, len, &result);
 
   *written = bytes_written;
-  if(bytes_written >= 0)
+  if(bytes_written >= 0) {
+    /* in the case of non-blocking sockets (e.g. TCP Fast Open) this
+       information may not have been available at the time of socket
```

Comment:
"this information". I'd just say it in one line if you could like `/* if local port was not available on first send (TFO), update conninfo */`

otherwise looks fine to me as long as local port once available !=0. is there any sockets library where that doesn't happen? let's see what @bagder and @ghedo think

Reply:
the only case where I can think of this happening would be with unix sockets or their NFS equivalent. I do seem to remember that cURL supports these maybe for proxies (I think I did tor testing with it - maybe it was a patch though).

Reply:
I think @jay's version of the comment is better as it explains the case better.

We might need some kind of precaution for non-TCP transfers to avoid calling that function many times. curl supports unix socket connections that are done instead of the TCP connection and for those, `local_port` will remain zero. Also, `Curl_updateconninfo` doesn't actually do anything for UDP transfers since there's no "connection" then.

Reply:
@bagder Would there be any case where primary_port is set but local_port would always be zero? For UDP connections, shouldn't local_port be set during the bind?